### ISSUE TITLE
[FEA] Support for pci and nvme commands in u-boot hardening

### ISSUE THREAD

**Issue by fboudra** [toradex/meta-toradex-security#75]
Hi,

The U-Boot hardening implementation lacks the support for pci and nvme commands.
 ```
## WARNING: Command execution denied (name not in whitelist) for `pci enum`.
## WARNING: Command execution denied (name not in whitelist) for `nvme scan`.
## WARNING: Command execution denied (name not in whitelist) for `nvme dev`.
 ```
 
 As a workaround, I added locally the categories and commands.
 
 Update:
 * Building Torizon OS 7 for verdin-imx8mp

**Comment by rborn-tx** [comment]
@fboudra Thanks for reporting.

Let me ask you:

- For what machine are you building?
- What kind of image are you building: Torizon OS, Toradex BSP or something else?

**Comment by fboudra** [comment]
I'm building Torizon OS 7 for verdin-imx8mp. For info, I have a NVMe SSD on PCIe on the platform. I can successfully boot with nvme_boot command once I closed the device and my local patch.

**Comment by rborn-tx** [comment]
@fboudra Thanks for the information.

I think it's perfectly acceptable for us to add the commands you mentioned to the whitelist. We're also considering making the default list of allowed/denied categories more easily configurable for Yocto users. We have internal discussions to decide when to do that though.

**Comment by rborn-tx** [comment]
@fboudra Would something like this work for you?


In `whitelist.c`:
```
@@ -305,6 +305,15 @@ static const struct cmd_whitelist_entry cmd_whitelist[] = {
 	{ { "net", "list", ALL }, { CMD_CAT_NET_DIAG } },
 	{ { "nfs", ALL }, { CMD_CAT_EXEC, CMD_CAT_MEM_WRITE } },
 	{ { "nm", ALL }, { CMD_CAT_MEM_READ, CMD_CAT_MEM_WRITE } },
+#ifdef CONFIG_CMD_NVME
+	{ { "nvme", "scan", ALL}, { CMD_CAT_NVME_CONTROL } },
+	{ { "nvme", "detail", ALL}, { CMD_CAT_NVME_DIAG } },
+	{ { "nvme", "info", ALL}, { CMD_CAT_NVME_DIAG } },
+	{ { "nvme", "device", ALL}, { CMD_CAT_NVME_CONTROL } },
+	{ { "nvme", "part", ALL}, { CMD_CAT_NVME_DIAG } },
+	{ { "nvme", "read", ALL}, { CMD_CAT_NVME_READ } },
+	{ { "nvme", "write", ALL}, { CMD_CAT_NVME_WRITE } },
+#endif /* CONFIG_CMD_NVME */
 	{ { "panic", ALL }, { CMD_CAT_SAFE } },
 	{ { "part", "uuid", ALL }, { CMD_CAT_PART_READ, CMD_CAT_NEEDED } },
 	{ { "part", "list", ALL }, { CMD_CAT_PART_READ, CMD_CAT_NEEDED } },
@@ -312,6 +321,27 @@ static const struct cmd_whitelist_entry cmd_whitelist[] = {
 	{ { "part", "size", ALL }, { CMD_CAT_PART_READ } },
 	{ { "part", "number", ALL }, { CMD_CAT_PART_READ } },
 	{ { "part", "types", ALL }, { CMD_CAT_PART_DIAG } },
+#ifdef CONFIG_CMD_PCI
+	{ { "pci", "enum", ALL }, { CMD_CAT_PCI_CONTROL } },
+	{ { "pci", "header", ALL }, { CMD_CAT_PCI_DIAG } },
+	{ { "pci", "bar", ALL }, { CMD_CAT_PCI_DIAG } },
+	{ { "pci", "regions", ALL }, { CMD_CAT_PCI_DIAG } },
+	{ { "pci", "display.b", ALL }, { CMD_CAT_PCI_READ } },
+	{ { "pci", "display.w", ALL }, { CMD_CAT_PCI_READ } },
+	{ { "pci", "display.l", ALL }, { CMD_CAT_PCI_READ } },
+	{ { "pci", "next.b", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "next.w", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "next.l", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "modify.b", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "modify.w", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "modify.l", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "write.b", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "write.w", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "write.l", ALL }, { CMD_CAT_PCI_WRITE } },
+	{ { "pci", "*" }, { CMD_CAT_PCI_DIAG } },
+	{ { "pci", "*", "long" }, { CMD_CAT_PCI_DIAG } },
+	{ { "pci", ALL }, { CMD_CAT_PCI_OTHER } },
+#endif /* CONFIG_CMD_PCI */
 	{ { "ping", ALL }, { CMD_CAT_NET_DIAG } },
 	{ { "pinmux", "list", ALL }, { CMD_CAT_PINMUX_DIAG } },
 	{ { "pinmux", "dev", ALL }, { CMD_CAT_PINMUX_DIAG } },
```

In `cmd-categories.h`:
```
@@ -74,8 +74,17 @@
 #define CMD_CAT_MMC_READ		252
 #define CMD_CAT_MMC_WRITE		253
 #define CMD_CAT_NET_DIAG		260
+#define CMD_CAT_NVME_CONTROL		265
+#define CMD_CAT_NVME_DIAG		266
+#define CMD_CAT_NVME_READ		267
+#define CMD_CAT_NVME_WRITE		268
 #define CMD_CAT_PART_DIAG		270
 #define CMD_CAT_PART_READ		271
+#define CMD_CAT_PCI_CONTROL		274
+#define CMD_CAT_PCI_DIAG		275
+#define CMD_CAT_PCI_READ		276
+#define CMD_CAT_PCI_WRITE		277
+#define CMD_CAT_PCI_OTHER		278
 #define CMD_CAT_PINMUX_DIAG		280
 #define CMD_CAT_PMIC_DIAG		290
 #define CMD_CAT_PMIC_READ		291
```

Of course you would need to add categories `CMD_CAT_NVME_CONTROL` and `CMD_CAT_PCI_CONTROL` to your allowed ~open and~ closed lists.

**Comment by fboudra** [comment]
Looks good to me. I'll test it locally tomorrow.

**Comment by rborn-tx** [comment]
@fboudra Btw, since [commit 2e873b1a9b45cc041b0a26c544224ce6a39412b0](https://github.com/toradex/meta-toradex-security/commit/2e873b1a9b45cc041b0a26c544224ce6a39412b0) one should be able to define the categories by just setting OE variables. In your case, it should be something like this:

```
TDX_SECBOOT_WL_ALLOW_CLOSED_CATEG = "CMD_CAT_ALL_SAFE CMD_CAT_NVME_CONTROL CMD_CAT_PCI_CONTROL" 
```

Notice I'm adding the command `nvme device` to the whitelist since this is the documented command in U-Boot even though based on the log messages you sent it seems you're using `nvme dev` instead. Checking the implementation in U-Boot I see anything starting with the character `d` would be mapped to the command `nvme device` but to pass the whitelist an exact match must be found.

I could also add `nvme dev` to the whitelist if it helps - please let me know if you see the need for that or if it's okay for you to change your boot script to use `nvme device` instead.

**Comment by fboudra** [comment]
fwiw, the default boot script on verdin-imx8mp has:
`nvme_boot=run boot_pci_enum; run nvme_init; if nvme dev ${devnum}; then devtype=nvme; run scan_dev_for_boot_part; fi`

hence it needs to be updated.

It's coming from https://source.denx.de/u-boot/u-boot/-/blob/master/include/config_distro_bootcmd.h?ref_type=heads#L29

```c
#define BOOTENV_SHARED_BLKDEV_BODY(devtypel) \
		"if " #devtypel " dev ${devnum}; then " \
			"devtype=" #devtypel "; " \
			"run scan_dev_for_boot_part; " \
		"fi\0"
```

It might be preferred to stick to "dev" instead of "device" to avoid updating the default distro boot script.

**Comment by fboudra** [comment]
Somehow related, meta-torizon has introduced a feature to fuse from u-boot. You get warnings for `fuse cmp` , `fuse readm` and `tdx_is_closed` commands. By default, those are producing warning because they aren't in allowlisting.

**Comment by fboudra** [comment]
@rborn-tx tested successfully with:

`TDX_SECBOOT_WL_ALLOW_CLOSED_CATEG = "CMD_CAT_ALL_SAFE CMD_CAT_FUSE_READ CMD_CAT_MMC_DIAG CMD_CAT_NVME_CONTROL CMD_CAT_PCI_CONTROL CMD_CAT_USB_CONTROL"`

My additional changes:

- "dev" instead of "device" ` +	{ { "nvme", "dev", ALL }, { CMD_CAT_NVME_CONTROL } },`

- added "tdx_is_closed" `+	{ { "tdx_is_closed", ALL }, { CMD_CAT_FUSE_READ } },`

**Comment by rborn-tx** [comment]
> It might be preferred to stick to "dev" instead of "device" to avoid updating the default distro boot script.

Agree. I'll leave both "device" and "dev" on the whitelist for compatibility.

> Somehow related, meta-torizon has introduced a feature to fuse from u-boot. You get warnings for fuse cmp , fuse readm and tdx_is_closed commands. By default, those are producing warning because they aren't in allowlisting.

Yes, I'm about to create a PR whose goal is exactly to adapt our U-Boot patches to the new fusing feature and since this involves reviewing the whitelist, I'm taking the opportunity to include the changes you asked for in the present issue.

Fyi, in this upcoming PR my intention is to mark `fuse cmp` as "needed", create a replacement to `fuse readm` (because this is a very insecure command that allows writes to arbitrary memory addresses) and to drop `tdx_is_close` in favor of a new more generic `tdx_secboot_get`).

> @rborn-tx tested successfully with:

Awesome! Thanks for the confirmation.

**Comment by fboudra** [comment]
Closing. Fixed in https://github.com/toradex/meta-toradex-security/pull/95
Tested on verdin-imx8mp

**Comment by fboudra** [comment]
@rborn-tx it looks like our CI is failing on qemuarm64:
```
| aarch64-tdx-linux-ld.bfd: common/tdx-secboot.o: in function `do_secboot_get':
| /usr/src/debug/u-boot/2024.01/common/tdx-secboot.c:219:(.text.do_secboot_get+0x60): undefined reference to `cmd_arg1'
| aarch64-tdx-linux-ld.bfd: /usr/src/debug/u-boot/2024.01/common/tdx-secboot.c:220:(.text.do_secboot_get+0x70): undefined reference to `cmd_arg2'
| make[1]: *** [/oe/build/tmp/work/qemuarm64-tdx-linux/u-boot/2024.01/git/Makefile:1765: u-boot] Error 139
| make[1]: *** Deleting file 'u-boot'
| make[1]: Leaving directory '/oe/build/tmp/work/qemuarm64-tdx-linux/u-boot/2024.01/build'
| make: *** [Makefile:177: sub-make] Error 2
| make: Leaving directory '/oe/build/tmp/work/qemuarm64-tdx-linux/u-boot/2024.01/git'
| ERROR: oe_runmake failed
```

**Comment by sergioprado** [comment]
@fboudra Be aware that `qemuarm64` is not in the list of supported machines for the secure boot feature. See https://github.com/toradex/meta-toradex-security/blob/kirkstone-6.x.y/docs/README-secure-boot.md.

If you are building for `qemuarm64`, I would recommend removing the `meta-toradex-security` layer from `BBLAYERS`.

**Comment by fboudra** [comment]
@sergioprado ack, make sense. It used to build at least fine with my manifest until now. Instead of removing the layer, I'll check if I can disable U-Boot hardening for the qemuarm64 target. Thanks.

**Comment by rborn-tx** [comment]
@fboudra Given the error, I see this is likely a side-effect of using `tdx-signed` with `qemuarm64` (which, as @sergioprado said, it's not supported). Since this new module `tdx-secboot` will be present independent of the hardening, I'd expect the failure to remain even if you disable the hardening. To actually prevent the use of that module you'd need to not use `tdx-signed`, at least in theory, but please let us know the actual results of your attempts.

**Comment by fboudra** [comment]
@rborn-tx you're right. I only updated my conf to limit tdx-signed to the supported machines:
`INHERIT:append:verdin-imx8mp = " tdx-signed"`

It worked fine. The build passed as expected. Thanks.
